File: java/90.subsets-ii.java
/*
 * @lc app=leetcode id=90 lang=java
 *
 * [90] Subsets II
 */

// @lc code=start
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Deque<Integer> q = new ArrayDeque();
        List<List<Integer>> res = new ArrayList();

        Arrays.sort(nums);

        buildSubset(nums, res, q, -1);

        return res;
    }

    public void buildSubset(int[] nums, List<List<Integer>> res, Deque<Integer> q, int i) {
        res.add(new ArrayList(q));

        for(int j = i + 1; j < nums.length; j++) {
            if(j > i + 1 && nums[j] == nums[j - 1]) {
                continue;
            }
            q.addLast(nums[j]);
            buildSubset(nums, res, q, j);
            q.removeLast();
        }
    }
}
// @lc code=end

File: cpp/90.subsets-ii.cpp
/*
 * @lc app=leetcode id=90 lang=cpp
 *
 * [90] Subsets II
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> curr;
        subset(res, nums, curr, 0);
        
        return res;
    }

    void subset(vector<vector<int>>& res, vector<int>& nums, vector<int>& curr, int i) {
        res.push_back(curr);

        while(i < nums.size()) {
            curr.push_back(nums[i]);
            subset(res, nums, curr, i + 1);
            curr.pop_back();

            do {
                i++;
            } while(i < nums.size() && nums[i] == nums[i-1]);
        }
    }
};
// @lc code=end

